switchapi/src/switch_packet.c

```diff
@@ -156,41 +183,198 @@ switch_packet_rx_from_hw()
 }
 
 void
-switch_packet_tx_to_host(switch_hostif_info_t *hostif_info, char *packet, int packet_size)
+switch_packet_rx_to_host(
+        switch_packet_header_t *packet_header,
+        char *packet,
+        int packet_size)
 {
-    int intf_fd = hostif_info->intf_fd;
-    if (write(intf_fd, packet, packet_size) < 0) {
+    switch_cpu_header_t               *cpu_header = NULL;
+    switch_packet_rx_info_t           *rx_info = NULL;
+    switch_ethernet_header_t          *eth_header = NULL;
+    switch_vlan_header_t              *vlan_header = NULL;
+    static char                        in_packet[SWITCH_PACKET_MAX_BUFFER_SIZE];
+    switch_packet_rx_entry_t           rx_entry;
+    int                                intf_fd = 0;
+    uint16_t                           offset = 0;
+    switch_status_t                    status = SWITCH_STATUS_SUCCESS;
+
+    cpu_header = &packet_header->cpu_header;
+
+    memset(&rx_entry, 0x0, sizeof(rx_entry));
+    rx_entry.port = cpu_header->ingress_port;
+    rx_entry.ifindex = cpu_header->ingress_ifindex;
+    rx_entry.bd = cpu_header->ingress_bd;
+    rx_entry.reason_code = cpu_header->reason_code;
+
+    status = switch_packet_rx_info_get(
+                             &rx_entry,
+                             &rx_info);
+    if (status != SWITCH_STATUS_SUCCESS) {
+        SWITCH_API_ERROR("failed to find fd. dropping packet");
+        return;
+    }
+
+    if (rx_info->vlan_action == SWITCH_PACKET_VLAN_ADD) {
+        memcpy(in_packet, packet, packet_size);
```

Comment:
performance - un-necessary pkt copy. Use else condition to copy the entire packet.